retriever/lib/scripts.py

```diff
@@ -150,59 +176,82 @@ def get_script_upstream(dataset, repo=REPOSITORY):
+    except requests.exceptions.ConnectionError:
+        return None
+
+
+def get_script_version_upstream(dataset, repo=REPOSITORY):
+    """Return the upstream script version for a named dataset."""
+    try:
+        script = dataset.replace('-', '_') + ".json"
+        filepath = "scripts/" + script
+        r = requests.get(repo + filepath, allow_redirects=True, stream=True)
+        if r.status_code == 404:
+            script = dataset.replace('-', '_') + ".py"
+            filepath = "scripts/" + script
+            r = requests.get(repo + filepath, allow_redirects=True, stream=True)
+            if r.status_code == 404:
+                return None
+
+        pattern = re.compile(r'[."]version[\'"\s":=]+(\d+\.\d+.\d+)')
+        version = re.search(pattern, r.text).group(1)
+        return version
+    except requests.exceptions.ConnectionError:
+        return None
 
 
 def get_dataset_names_upstream(keywords=None, licenses=None, repo=REPOSITORY):
     """Search all datasets upstream by keywords and licenses."""
-    if not keywords and not licenses:
-        version_file = requests.get(repo + "version.txt").text
-        version_file = version_file.splitlines()[1:]
-
-        scripts = []
-        max_scripts = 100
-        for line in version_file:
-            script = line.strip('\n').split(',')[0]
-            script = '.'.join(script.split('.')[:-1])
-            script = script.replace('_', '-')
-            scripts.append(script)
-            if len(scripts) == max_scripts:
-                break
-        return sorted(scripts)
-
-    result_scripts = set()
-    if repo == RETRIEVER_REPOSITORY:
-        search_url = "https://api.github.com/search/code?q={query}+in:file+path:scripts+repo:weecology/retriever"
-    else:
-        search_url = "https://api.github.com/search/code?q={query}+in:file+path:scripts+repo:weecology/retriever-recipes"
-    if licenses:
-        licenses = [l.lower() for l in licenses]
-        for l in licenses:
-            try:
-                r = requests.get(search_url.format(query=l))
-                r = r.json()
-                for index in range(r['total_count']):
-                    script = r['items'][index]['name']
-                    script = '.'.join(script.split('.')[:-1])
-                    script = script.replace('_', '-')
-                    result_scripts.add(script)
-            except:
-                raise
-    if keywords:
-        keywords = [k.lower() for k in keywords]
-        for k in keywords:
-            try:
-                r = requests.get(search_url.format(query=k))
-                r = r.json()
-                for index in range(r['total_count']):
-                    script = r['items'][index]['name']
-                    script = '.'.join(script.split('.')[:-1])
-                    script = script.replace('_', '-')
-                    result_scripts.add(script)
-            except:
-                raise
-    return sorted(result_scripts)
+    try:
```

Comment:
If you want to catch requests exceptions, I think you should catch the base class (requests.exceptions.RequestException). There are a bunch of requests errors that might slip through this check (HTTPError, TimeoutError). 

Now that we are switching over to calls and searches to things upstream, I think this is a good idea to add try catches. I think a better design though would be to abstract retriever specific calls in a new function that solely deals with calls upstream and returning information. This way when something goes wrong with requests, we can report it in retriever without having to try/catch every piece of retriever making calls upstream. The new function could be leveraged by get_dataset_names_upstream and also get_script_version_upstream

Reply:
@zhangcandrew sorry, your idea is great but could we please do this later in a separate pr?

Reply:
Sure if you're more comfortable with that. I think you can add in the exception change relatively quick, and we can make a note about tackling this in a different PR.